std/algorithm/sorting.d

```diff
@@ -1048,147 +1048,287 @@ unittest
             (index3));
 }
 
-/**
-Lazily computes the union of two or more ranges $(D rs). The ranges
-are assumed to be sorted by $(D less). Elements in the output are not
-unique; the length of the output is the sum of the lengths of the
-inputs. (The $(D length) member is offered if all ranges also have
-length.) The element types of all ranges must have a common type.
-
-Params:
-    less = Predicate the given ranges are sorted by.
-    rs = The ranges to compute the union for.
-
-Returns:
-    A range containing the union of the given ranges.
-
-See_Also:
-   $(XREF algorithm, setops, SetUnion)
- */
-struct Merge(alias less = "a < b", Rs...) if (allSatisfy!(isInputRange, Rs))
+struct Merge(alias less = "a < b", Rs...)
+    if (Rs.length >= 2 &&
+        allSatisfy!(isInputRange, Rs) &&
+        is(CommonType!(staticMap!(ElementType, Rs))))
 {
-private:
-    Rs _r;
-    alias comp = binaryFun!(less);
-    uint _crt;
+    alias ElementType = CommonType!(staticMap!(.ElementType, Rs));
+    enum isBidirectional = allSatisfy!(isBidirectionalRange, staticMap!(Unqual, Rs));
+    this(Rs source)
+    {
+        this.source = source;
+        this._lastFrontIndex = frontIndex;
+        static if (isBidirectional)
+            this._lastBackIndex = backIndex;
```

Comment:
This may be problematic. A vast fraction of `Merge` uses is in an input/forward manner, even though the inputs are bidirectional. Therefore computing eagerly matters related to backwards iteration is often adding overhead. Can you arrange computation such that any work related to backward iteration is done lazily?